chips/sam4l/src/spi.rs

```diff
@@ -12,64 +12,144 @@ use dma::DMAChannel;
 use dma::DMAClient;
 use dma::DMAPeripheral;
 use kernel::ReturnCode;
-use kernel::common::VolatileCell;
+use kernel::common::regs::{ReadOnly, ReadWrite, WriteOnly};
 use kernel::hil::spi;
 use kernel::hil::spi::ClockPhase;
 use kernel::hil::spi::ClockPolarity;
 use kernel::hil::spi::SpiMasterClient;
 use kernel::hil::spi::SpiSlaveClient;
 use pm;
 
-/// The registers used to interface with the hardware
-#[repr(C)]
-struct SpiRegisters {
-    cr: VolatileCell<u32>,       //       0x0
-    mr: VolatileCell<u32>,       //       0x4
-    rdr: VolatileCell<u32>,      //      0x8
-    tdr: VolatileCell<u32>,      //      0xC
-    sr: VolatileCell<u32>,       //       0x10
-    ier: VolatileCell<u32>,      //      0x14
-    idr: VolatileCell<u32>,      //      0x18
-    imr: VolatileCell<u32>,      //      0x1C
-    _reserved0: [u32; 4],        //        0x20, 0x24, 0x28, 0x2C
-    csr0: VolatileCell<u32>,     //     0x30
-    csr1: VolatileCell<u32>,     //     0x34
-    csr2: VolatileCell<u32>,     //     0x38
-    csr3: VolatileCell<u32>,     //     0x3C
-    _reserved1: [u32; 41],       //       0x40 - 0xE0
-    wpcr: VolatileCell<u32>,     //     0xE4
-    wpsr: VolatileCell<u32>,     //     0xE8
-    _reserved2: [u32; 3],        //        0xEC - 0xF4
-    features: VolatileCell<u32>, // 0xF8
-    version: VolatileCell<u32>,  //  0xFC
+#[repr(C, packed)]
+pub struct Registers {
+    pub cr: WriteOnly<u32, Control::Register>,
```

Comment:
This is probably the right time to ask the question as to whether registers should be named to match the datasheet or to be human readable. I know we've discussed it before, but I don't know if we ever made a decision.

Reply:
I'm ambivilant... Human readable names means code ends up being harder to follow along-side the datasheet, datasheet names make it harder to understand code otherwise...

Reply:
Might there be a way to do both? I.e. make one an alias of the other
(I'm not seeing how this could be done myself, but it's certainly worth thinking about)

Reply:
I doubt there is a solution to this. Probably not worth covering on every chip peripheral PR.

Reply:
After thinking about it for a few more minutes, I'm changing my vote to using datasheet names, since it's (a) what Shane and you already did and (b) it's unambiguous... if we use datasheet names, we can always just go to the datasheet to figure out what they do, if we make up "human readable" ones, there's a chance we choose bad names and then we have the worst of both worlds.

These are low enough in the stack that I maybe chip-specific jargon is fine.

Reply:
@alevy great argument